File: drivers/nodejs/boyodb/subscriptions.js
/**
 * Pub/Sub and CDC subscription support for BoyoDB Node.js driver.
 *
 * Provides real-time messaging and change data capture subscriptions.
 */

/**
 * Change types for CDC events.
 * @enum {string}
 */
const ChangeType = {
  INSERT: 'INSERT',
  UPDATE: 'UPDATE',
  DELETE: 'DELETE',
};

/**
 * Message received from a subscription.
 * @typedef {Object} Message
 * @property {string} channel - Channel name
 * @property {*} payload - Message payload
 * @property {Date} timestamp - Message timestamp
 * @property {string} [id] - Message ID
 */

/**
 * CDC change event.
 * @typedef {Object} ChangeEvent
 * @property {string} table - Table name
 * @property {ChangeType} changeType - Type of change
 * @property {Object} [before] - Row before change (for UPDATE/DELETE)
 * @property {Object} after - Row after change (for INSERT/UPDATE)
 * @property {number} lsn - Log sequence number
 * @property {Date} timestamp - Change timestamp
 */

/**
 * Pub/Sub subscriber for real-time messaging.
 *
 * @example
 * const subscriber = new Subscriber(client, ['events', 'alerts'], (msg) => {
 *   console.log(`[${msg.channel}] ${JSON.stringify(msg.payload)}`);
 * });
 *
 * subscriber.start();
 *
 * // Later...
 * subscriber.stop();
 */
class Subscriber {
  /**
   * Create a new subscriber.
   * @param {Client} client - BoyoDB client
   * @param {string[]} channels - Channels to subscribe to
   * @param {Function} onMessage - Callback for received messages
   * @param {Object} [options={}] - Subscriber options
   * @param {Function} [options.onError] - Error callback
   * @param {number} [options.reconnectDelayMs=1000] - Reconnect delay
   */
  constructor(client, channels, onMessage, options = {}) {
    this._client = client;
    this._channels = channels;
    this._onMessage = onMessage;
    this._onError = options.onError || console.error;
    this._reconnectDelayMs = options.reconnectDelayMs || 1000;

    this._running = false;
    this._pollTimer = null;
  }

  /**
   * Start the subscriber.
   */
  start() {
    if (this._running) return;
    this._running = true;

    // Subscribe to channels
    for (const channel of this._channels) {
      this._subscribe(channel).catch(this._onError);
    }

    this._startPolling();
  }

  /**
   * Stop the subscriber.
   */
  stop() {
    this._running = false;
    if (this._pollTimer) {
      clearTimeout(this._pollTimer);
      this._pollTimer = null;
    }

    // Unsubscribe from channels
    for (const channel of this._channels) {
      this._unsubscribe(channel).catch(this._onError);
    }
  }

  /**
   * Add a channel to subscribe to.
   * @param {string} channel - Channel name
   */
  async addChannel(channel) {
    if (!this._channels.includes(channel)) {
      this._channels.push(channel);
      if (this._running) {
        await this._subscribe(channel);
      }
    }
  }

  /**
   * Remove a channel subscription.
   * @param {string} channel - Channel name
   */
  async removeChannel(channel) {
    const idx = this._channels.indexOf(channel);
    if (idx !== -1) {
      this._channels.splice(idx, 1);
      if (this._running) {
        await this._unsubscribe(channel);
      }
    }
  }

  async _subscribe(channel) {
    await this._client.exec(`SUBSCRIBE '${channel}'`);
  }

  async _unsubscribe(channel) {
    await this._client.exec(`UNSUBSCRIBE '${channel}'`);
  }

  _startPolling() {
    if (!this._running) return;

    this._pollTimer = setTimeout(async () => {
      try {
        const result = await this._client.query('POLL MESSAGES');
        for (const row of result) {
          const msg = {
            channel: row.channel,
            payload: typeof row.payload === 'string' ? JSON.parse(row.payload) : row.payload,
            timestamp: new Date(row.timestamp),
            id: row.id,
          };
          this._onMessage(msg);
        }
      } catch (err) {
        this._onError(err);
      }
      this._startPolling();
    }, 100);
  }
}

/**
 * CDC subscriber for change data capture.
 *
 * @example
 * const cdc = new CDCSubscriber(client, 'mydb.users', (event) => {
 *   console.log(`${event.changeType}: ${JSON.stringify(event.after)}`);
 * }, {
 *   includeBefore: true,
 *   changeTypes: [ChangeType.INSERT, ChangeType.UPDATE]
 * });
 *
 * cdc.start();
 *
 * // Later...
 * cdc.stop();
 */
class CDCSubscriber {
  /**
   * Create a new CDC subscriber.
   * @param {Client} client - BoyoDB client
   * @param {string} table - Table to watch (database.table)
   * @param {Function} onChange - Callback for change events
   * @param {Object} [options={}] - Subscriber options
   * @param {boolean} [options.includeBefore=false] - Include before image
   * @param {ChangeType[]} [options.changeTypes] - Filter by change types
   * @param {Function} [options.onError] - Error callback
   * @param {number} [options.startLsn] - Starting LSN
   */
  constructor(client, table, onChange, options = {}) {
    this._client = client;
    this._table = table;
    this._onChange = onChange;
    this._includeBefore = options.includeBefore || false;
    this._changeTypes = options.changeTypes || Object.values(ChangeType);
    this._onError = options.onError || console.error;
    this._startLsn = options.startLsn || 0;

    this._running = false;
    this._currentLsn = this._startLsn;
    this._pollTimer = null;
  }

  /**
   * Start the CDC subscriber.
   */
  start() {
    if (this._running) return;
    this._running = true;

    this._startCDC().catch(this._onError);
    this._startPolling();
  }

  /**
   * Stop the CDC subscriber.
   */
  stop() {
    this._running = false;
    if (this._pollTimer) {
      clearTimeout(this._pollTimer);
      this._pollTimer = null;
    }
    this._stopCDC().catch(this._onError);
  }

  /**
   * Get current LSN position.
   * @returns {number}
   */
  get currentLsn() {
    return this._currentLsn;
  }

  async _startCDC() {
    const opts = [];
    if (this._includeBefore) {
      opts.push("include_before = true");
    }
    if (this._startLsn > 0) {
      opts.push(`start_lsn = ${this._startLsn}`);
    }

    const optStr = opts.length > 0 ? ` WITH (${opts.join(', ')})` : '';
    await this._client.exec(`START CDC ON ${this._table}${optStr}`);
  }

  async _stopCDC() {
    await this._client.exec(`STOP CDC ON ${this._table}`);
  }

  _startPolling() {
    if (!this._running) return;

    this._pollTimer = setTimeout(async () => {
      try {
        const result = await this._client.query(
          `POLL CDC ${this._table} FROM ${this._currentLsn}`
        );

        for (const row of result) {
          const changeType = row.change_type || row.changeType;

          // Filter by change type
          if (!this._changeTypes.includes(changeType)) {
            continue;
          }

          const event = {
            table: this._table,
            changeType: changeType,
            before: row.before,
            after: row.after,
            lsn: row.lsn,
            timestamp: new Date(row.timestamp),
          };

          this._onChange(event);
          this._currentLsn = Math.max(this._currentLsn, row.lsn + 1);
        }
      } catch (err) {
        this._onError(err);
      }
      this._startPolling();
    }, 100);
  }
}

/**
 * Publisher for pub/sub messaging.
 *
 * @example
 * const publisher = new Publisher(client);
 * await publisher.publish('events', { type: 'user_signup', userId: 123 });
 * await publisher.publish('alerts', { level: 'warning', message: 'High CPU' });
 */
class Publisher {
  /**
   * Create a new publisher.
   * @param {Client} client - BoyoDB client
   */
  constructor(client) {
    this._client = client;
  }

  /**
   * Publish a message to a channel.
   * @param {string} channel - Channel name
   * @param {*} payload - Message payload
   * @returns {Promise<void>}
   */
  async publish(channel, payload) {
    const payloadStr = typeof payload === 'string'
      ? payload
      : JSON.stringify(payload);

    const escaped = payloadStr.replace(/'/g, "''");
    await this._client.exec(`PUBLISH '${channel}' '${escaped}'`);
  }

  /**
   * Publish multiple messages to a channel.
   * @param {string} channel - Channel name
   * @param {*[]} payloads - Array of payloads
   * @returns {Promise<void>}
   */
  async publishMany(channel, payloads) {
    for (const payload of payloads) {
      await this.publish(channel, payload);
    }
  }
}

module.exports = {
  ChangeType,
  Subscriber,
  CDCSubscriber,
  Publisher,
};

File: drivers/python/boyodb/subscriptions.py
"""
Pub/Sub and CDC subscriptions for BoyoDB.

Provides real-time data streaming capabilities.
"""

import json
import threading
import time
from dataclasses import dataclass
from enum import Enum
from typing import Any, Callable, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .client import Client


class ChangeType(Enum):
    """CDC change types."""
    INSERT = "insert"
    UPDATE = "update"
    DELETE = "delete"
    TRUNCATE = "truncate"


@dataclass
class ChangeEvent:
    """CDC change event."""
    table: str
    change_type: ChangeType
    before: Optional[Dict[str, Any]]  # For UPDATE/DELETE
    after: Optional[Dict[str, Any]]   # For INSERT/UPDATE
    timestamp: int
    transaction_id: Optional[str]
    sequence: int


@dataclass
class Message:
    """Pub/Sub message."""
    channel: str
    payload: Any
    timestamp: int


class Subscriber:
    """
    Pub/Sub subscriber for real-time messaging.

    Usage:
        def on_message(msg):
            print(f"Received: {msg.payload}")

        sub = Subscriber(client, ["events", "alerts"], on_message)
        sub.start()
        # ...
        sub.stop()
    """

    def __init__(
        self,
        client: "Client",
        channels: List[str],
        callback: Callable[[Message], None],
        error_callback: Optional[Callable[[Exception], None]] = None,
    ):
        self._client = client
        self._channels = channels
        self._callback = callback
        self._error_callback = error_callback
        self._running = False
        self._thread: Optional[threading.Thread] = None

    def start(self) -> None:
        """Start listening for messages."""
        if self._running:
            return

        # Subscribe to channels
        for channel in self._channels:
            self._client.exec(f"LISTEN {channel}")

        self._running = True
        self._thread = threading.Thread(target=self._listen_loop, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        """Stop listening and unsubscribe."""
        self._running = False

        # Unsubscribe from channels
        for channel in self._channels:
            try:
                self._client.exec(f"UNLISTEN {channel}")
            except Exception:
                pass

        if self._thread:
            self._thread.join(timeout=5.0)

    def _listen_loop(self) -> None:
        """Background listening loop."""
        while self._running:
            try:
                # Poll for notifications
                result = self._client.query("SELECT pg_notification_poll()")
                for row in result:
                    if row.get("channel") and row.get("payload"):
                        msg = Message(
                            channel=row["channel"],
                            payload=self._parse_payload(row["payload"]),
                            timestamp=row.get("timestamp", int(time.time() * 1000)),
                        )
                        self._callback(msg)

            except Exception as e:
                if self._error_callback:
                    self._error_callback(e)

            time.sleep(0.1)  # Poll interval

    @staticmethod
    def _parse_payload(payload: str) -> Any:
        """Try to parse payload as JSON."""
        try:
            return json.loads(payload)
        except (json.JSONDecodeError, TypeError):
            return payload


class CDCSubscriber:
    """
    CDC (Change Data Capture) subscriber for table changes.

    Usage:
        def on_change(event):
            print(f"{event.change_type}: {event.after}")

        cdc = CDCSubscriber(client, "mydb.users", on_change)
        cdc.start()
        # ...
        cdc.stop()
    """

    def __init__(
        self,
        client: "Client",
        table: str,
        callback: Callable[[ChangeEvent], None],
        error_callback: Optional[Callable[[Exception], None]] = None,
        include_before: bool = True,
        change_types: Optional[List[ChangeType]] = None,
    ):
        self._client = client
        self._table = table
        self._callback = callback
        self._error_callback = error_callback
        self._include_before = include_before
        self._change_types = change_types or list(ChangeType)
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._last_sequence = 0
        self._slot_name: Optional[str] = None

    def start(self) -> None:
        """Start receiving CDC events."""
        if self._running:
            return

        # Create replication slot
        self._slot_name = f"cdc_{self._table.replace('.', '_')}_{int(time.time())}"

        try:
            self._client.exec(
                f"SELECT pg_create_logical_replication_slot('{self._slot_name}', 'boyodb_cdc')"
            )
        except Exception:
            # Slot might already exist or be managed differently
            pass

        self._running = True
        self._thread = threading.Thread(target=self._poll_loop, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        """Stop receiving CDC events."""
        self._running = False

        # Drop replication slot
        if self._slot_name:
            try:
                self._client.exec(
                    f"SELECT pg_drop_replication_slot('{self._slot_name}')"
                )
            except Exception:
                pass

        if self._thread:
            self._thread.join(timeout=5.0)

    def _poll_loop(self) -> None:
        """Background polling loop."""
        while self._running:
            try:
                # Query for changes
                result = self._client.query(
                    f"""
                    SELECT * FROM pg_logical_slot_get_changes(
                        '{self._slot_name}',
                        NULL,
                        1000,
                        'table', '{self._table}',
                        'include_before', '{str(self._include_before).lower()}'
                    )
                    """
                )

                for row in result:
                    event = self._parse_change(row)
                    if event and event.change_type in self._change_types:
                        self._callback(event)
                        self._last_sequence = event.sequence

            except Exception as e:
                if self._error_callback:
                    self._error_callback(e)

            time.sleep(0.1)  # Poll interval

    def _parse_change(self, row: Dict[str, Any]) -> Optional[ChangeEvent]:
        """Parse a change row into a ChangeEvent."""
        try:
            data = row.get("data") or row
            if isinstance(data, str):
                data = json.loads(data)

            change_type_str = data.get("op", data.get("change_type", "insert"))
            change_type = ChangeType(change_type_str.lower())

            return ChangeEvent(
                table=data.get("table", self._table),
                change_type=change_type,
                before=data.get("before"),
                after=data.get("after"),
                timestamp=data.get("ts", int(time.time() * 1000)),
                transaction_id=data.get("txid"),
                sequence=data.get("seq", self._last_sequence + 1),
            )
        except Exception:
            return None


class Publisher:
    """
    Pub/Sub publisher for sending messages.

    Usage:
        pub = Publisher(client)
        pub.publish("events", {"type": "user_signup", "user_id": 123})
    """

    def __init__(self, client: "Client"):
        self._client = client

    def publish(self, channel: str, payload: Any) -> None:
        """Publish a message to a channel."""
        if isinstance(payload, (dict, list)):
            payload_str = json.dumps(payload).replace("'", "''")
        else:
            payload_str = str(payload).replace("'", "''")

        self._client.exec(f"NOTIFY {channel}, '{payload_str}'")

    def publish_many(self, channel: str, payloads: List[Any]) -> None:
        """Publish multiple messages to a channel."""
        for payload in payloads:
            self.publish(channel, payload)
